trin-core/src/portalnet/find/query_info.rs

```diff
@@ -0,0 +1,135 @@
+use discv5::{kbucket::Key, Enr};
+use crate::portalnet::types::messages::{Request, FindNodes};
+use discv5::enr::NodeId;
+use sha3::digest::generic_array::GenericArray;
+use smallvec::SmallVec;
+use tokio::sync::oneshot;
+use crate::portalnet::find::query_pool::TargetKey;
+
+/// Information about a query.
+#[derive(Debug)]
+pub struct QueryInfo {
+    /// What we are querying and why.
+    pub query_type: QueryType,
+
+    /// Temporary ENRs used when trying to reach nodes.
+    pub untrusted_enrs: SmallVec<[Enr; 16]>,
+
+    /// The number of distances we request for each peer.
+    pub distances_to_request: usize,
+}
+
+/// Additional information about the query.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum QueryType {
+    /// The user requested a `FIND_NODE` query to be performed. It should be reported when finished.
+    FindNode(NodeId),
+}
+
+impl QueryInfo {
+    /// Builds an RPC Request, given the QueryInfo
+    pub(crate) fn rpc_request(&self, peer: NodeId) -> Result<Request, &'static str> {
+        let request = match self.query_type {
+            QueryType::FindNode(node_id) => {
+                let distances = findnode_log2distance(node_id, peer, self.distances_to_request)
+                    .ok_or("Requested a node find itself")?;
+                Request::FindNodes(FindNodes { distances })
+            }
+        };
+
+        Ok(request)
+    }
+}
+
+impl TargetKey<NodeId> for QueryInfo {
+    fn key(&self) -> Key<NodeId> {
+        match self.query_type {
+            QueryType::FindNode(ref node_id) => {
+                Key::new_raw(*node_id, *GenericArray::from_slice(&node_id.raw()))
+            }
+        }
+    }
+}
+
+/// Calculates the log2 distance for a destination peer given a target and the size (number of
+/// distances to request).
+///
+/// As the iteration increases, FINDNODE requests adjacent distances from the exact peer distance.
+///
+/// As an example, if the target has a distance of 12 from the remote peer, the sequence of distances that are sent for increasing iterations would be [12, 13, 11, 14, 10, .. ].
+fn findnode_log2distance(target: NodeId, peer: NodeId, size: usize) -> Option<Vec<u16>> {
+    if size > 127 {
+        // invoke and endless loop - coding error
+        panic!("Iterations cannot be greater than 127");
+    }
+
+    let dst_key: Key<NodeId> = peer.into();
+    let distance_u64 = dst_key.log2_distance(&target.into())?;
```

Comment:
The `log2_distance` method relies on [XOR distance](https://github.com/sigp/discv5/blob/master/src/kbucket/key.rs#L91) defined on `Key` in `sigp/discv5`. This is not a problem for the history network, but this may require some rework for the state network.

Reply:
good point, i'll file an issue after i merge this